### PR TITLE
Galaxy WL reformulation

### PR DESCRIPTION
**Pull request by caioolivv**
New galaxy weak lensing mass estimate model with the following structure:

- `NcGalaxyWLLikelihood` object which contains the following objects:
  - `NcGalaxySampleDistributionPosition` with information on source position and redshift;
  - `NcGalaxySampleDistributionShape` with information on intrinsic shape, observed shape and the shear mapping;
  - `NcGalaxySampleDistributionZProxy` with information on proxy redshift;

### CODE REVIEW

**Comment by caioolivv** [comment]
There are some problems to solve still:

1. I had to name new object `nc_galaxy_wl_likelihood` in order to not mess with previous naming scheme. I don't think that's the best name.
2. `nc_galaxy_wl_likelihood` currently has 2 methods `eval_m2lnP` and `kde_eval_m2lnP` for each of the methods. I don't know if that's the best way to present these methods. We can use a flag maybe?
3. I still have to implement the likelihood algorithms, both for KDE and integration.
4. If we wish to use different source positions priors and descriptions (i.e. using coordinates, $r$ or $\theta$) it might be best to make `nc_galaxy_sd_position` contain other two objects describing image position and redshift separately. This may not be necessary if we're not dealing with miscentering.
5. I think it's worth considering adding a way to deal with shape measurement error. Maybe by creating an optional object `nc_galaxy_sd_shape_m_err` witch could be contained by `nc_galaxy_sd_shape`.

Also, I'm already working on unit testing. That should be up today still.

**Comment by caioolivv** [comment]
I can't create unit tests yet since I haven't implemented `nc_galaxy_sd_*_new ()` and thus can't create neither `nc_galaxy_sd_*` objects nor `nc_galaxy_wl_likelihood` objects by extension. Maybe we merge the PR without unit testing while I work on the implementation of basic `nc_galaxy_sd_*` objects?

Also, I need some directions on how to write these new objects on `numcosmo-docs.sgml`.

File: numcosmo/data/nc_data_cluster_wll.c [hunk 1]
```diff
@@ -134,48 +226,187 @@ nc_data_cluster_wll_get_property (GObject *object, guint prop_id, GValue *value,
...
   /* Chain up : end */
-  G_OBJECT_CLASS (nc_data_cluster_wll_parent_class)->finalize (object);
+  G_OBJECT_CLASS (nc_data_cluster_wl_parent_class)->finalize (object);
 }
 
-static void _nc_data_cluster_wll_m2lnL_val (NcmData *data, NcmMSet *mset, gdouble *m2lnL);
-static guint _nc_data_cluster_wll_get_len (NcmData *data);
-static void _nc_data_cluster_wll_prepare (NcmData *data, NcmMSet *mset);
+static void
+nc_data_cluster_wl_constructed (GObject *object)
```
Comment (caioolivv) [inline numcosmo/data/nc_data_cluster_wll.c]: Rename to _nc_data_cluster_wl_constructed

File: numcosmo/data/nc_data_cluster_wll.c [hunk 2]
```diff
@@ -185,6 +416,13 @@ nc_data_cluster_wll_class_init (NcDataClusterWLLClass *klass)
...
+  /**
+   * NcDataClusterWL:kde-method:
+   *
+   * Whether to use KDE method.
+   *
+   */
+
   g_object_class_install_property (object_class,
                                    PROP_KDE,
                                    g_param_spec_boolean ("kde",
```
Comment (caioolivv) [inline numcosmo/data/nc_data_cluster_wll.c]: change "kde" to "use-kde"

File: numcosmo/data/nc_data_cluster_wl.c [hunk 3]
```diff
@@ -536,29 +537,139 @@ nc_data_cluster_wl_prepare_kde (NcDataClusterWL *dcwl, NcHICosmo *cosmo, NcHaloD
...
-  NcWLSurfaceMassDensity *smd         = NC_WL_SURFACE_MASS_DENSITY (ncm_mset_peek (mset, nc_wl_surface_mass_density_id ()));
-  NcHaloDensityProfile *dp            = NC_HALO_DENSITY_PROFILE (ncm_mset_peek (mset, nc_halo_density_profile_id ()));
-  guint i;
+  gdouble verr, vzpi, vzpf, vres, vtmp;
+  NcmVector *err  = ncm_vector_new_data_static (&verr, 1, 1);
+  NcmVector *zpi  = ncm_vector_new_data_static (&vzpi, 1, 1);
+  NcmVector *zpf  = ncm_vector_new_data_static (&vzpf, 1, 1);
+  NcmVector *res  = ncm_vector_new_data_static (&vres, 1, 1);
+  NcmVector *tmp  = ncm_vector_new_data_static (&vtmp, 1, 1);
+  const guint len = ncm_matrix_nrows (self->obs);
```
Comment (caioolivv) [inline numcosmo/data/nc_data_cluster_wl.c:564]: Use self->len instead

File: numcosmo/data/nc_data_cluster_wll.c [hunk 4]
```diff
@@ -194,168 +432,478 @@ nc_data_cluster_wll_class_init (NcDataClusterWLLClass *klass)
...
-  for (i = 0; i < ngal; i++)
+static void
+nc_data_cluster_wl_int_dim (NcmIntegralND *intnd, guint *dim, guint *fdim)
+{
+  *dim  = 1;
+  *fdim = 1;
+}
+
+void
+nc_data_cluster_wl_prepare_kde (NcDataClusterWL *dcwl, NcHICosmo *cosmo, NcHaloDensityProfile *dp, NcWLSurfaceMassDensity *smd, const gdouble z_cluster)
```
Comment (caioolivv) [inline numcosmo/data/nc_data_cluster_wll.c]: z_cluster doesn't need to be an argument

File: numcosmo/data/nc_data_cluster_wl.c [hunk 5]
```diff
@@ -575,93 +686,73 @@ _nc_data_cluster_wl_m2lnL_val (NcmData *data, NcmMSet *mset, gdouble *m2lnL)
...
+  if (self->kde_method)
+    m2lnL[0] += nc_data_cluster_wl_kde_eval_m2lnP (dcwl, cosmo, dp, smd, NULL);
+  else
+    m2lnL[0] += nc_data_cluster_wl_eval_m2lnP (dcwl, cosmo, dp, smd, NULL);
 
   return;
 }
 
+static guint
+_nc_data_cluster_wl_get_len (NcmData *data)
```
Comment (caioolivv) [inline numcosmo/data/nc_data_cluster_wl.c:748]: Check if obs has been set

File: numcosmo/data/nc_data_cluster_wl.c [hunk 6]
```diff
@@ -690,13 +781,13 @@ _nc_data_cluster_wl_prepare (NcmData *data, NcmMSet *mset)
 NcDataClusterWL *
 nc_data_cluster_wl_new (NcGalaxySDShape *s_dist, NcGalaxySDZProxy *zp_dist, NcGalaxySDPosition *rz_dist)
 {
-  NcDataClusterWL *gwl = g_object_new (NC_TYPE_DATA_CLUSTER_WL,
-                                       "s-dist", s_dist,
-                                       "zp-dist", zp_dist,
-                                       "rz-dist", rz_dist,
-                                       NULL);
+  NcDataClusterWL *dcwl = g_object_new (NC_TYPE_DATA_CLUSTER_WL,
```
Comment (caioolivv) [inline numcosmo/data/nc_data_cluster_wl.c:784]: Require use_kde and z_cluster

**Comment by caioolivv** [comment]
Setting the integral precision to 10^{-14} or lower will fail the integral with error `NUMCOSMO:ERROR:../numcosmo/math/ncm_integral_nd.c:625:ncm_integral_nd_eval: assertion failed (ret == 0): (1 == 0)`.

**Review by vitenti — APPROVED** [review]

**Pull request merged**